myia/anf_ir.py

```diff
@@ -0,0 +1,304 @@
+"""Intermediate representation definition.
+
+Myia's main intermediate representation (IR) is a graph-based version of ANF.
+Each function definition (lambda) is defined as a graph, consisting of a series
+of function applications.
+
+A function can be applied to a node from another funtion's graph; this
+implicitly creates a nested function. Functions are first-class objects, so
+returning a nested function creates a closure.
+
+"""
+import types
+from ast import AST
+from typing import (List, Set, Tuple, Any, Sequence, MutableSequence,
+                    overload, Iterable)
+
+from myia.ir import Node
+from myia.utils import Named
+
+PARAMETER = Named('PARAMETER')
+
+
+class Graph:
+    """A function graph.
+
+    Attributes:
+        parameters: The parameters of this function as a list of `Parameter`
+            nodes. Parameter nodes that are unreachable by walking from the
+            output node correspond to unused parameters.
+        return_: The `Apply` node that calls the `Return` primitive. The input
+            to this node will be returned by this function. A graph initially
+            has no output node (because it won't be known e.g. until the
+            function has completed parsing), but it must be set afterwards for
+            the graph instance to be valid.
+
+    """
+
+    def __init__(self) -> None:
+        """Construct a graph."""
+        self.parameters: List[Parameter] = []
+        self.return_: Apply = None
```

Comment:
`return_` bothers me a little, aesthetically. `ret`?

Reply:
The trailing underscore is the convention in Python when avoiding conflicts with keywords though, see https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles. In that context, I find `return_` a bit less ambiguous than `ret`, which might not make as much sense if you don't speak x86 assembly 😛

Reply:
Sure, I just think that's really ugly. I suppose I can stomach it.